### PR TITLE
Support for sd Quattro H

### PR DESCRIPTION
**Pull request by rolkar**
Adds support for sd Quattro H

Issue #96

### CODE REVIEW

File: src/x3f_process.c [hunk 1]
```diff
@@ -364,14 +365,19 @@ static void interpolate_bad_pixels(x3f_t *x3f, x3f_area16_t *image, int colors)
...
      TODO: The positions shouldn't really be hardcoded. */
-  #define SDQ_CAMERAID 40
-  if (x3f_get_camf_unsigned(x3f, "CAMERAID", &cameraid) &&
-      cameraid == SDQ_CAMERAID) {
-    static const grid_t sdq_af_luma = {217, 5641, 16, 1, 464, 3312, 32, 2};
-    static const grid_t sdq_af_chroma = {108, 2820, 8, 1, 232, 1656, 16, 1};
-    const grid_t *g = colors == 1 ? &sdq_af_luma : &sdq_af_chroma;
+  
+  x3f_get_camf_unsigned(x3f, "CAMERAID", &cameraid);
+
```
Comment (mmroden) [inline src/x3f_process.c]: why not two if statements?  (this is a genuine style question, the code is fine).

**Comment by mmroden** [comment]
Just to be clear, this doesn't fix issues with the new firmware for the quattros, or with the dark levels from the top/bottom vs left/right, correct?  Just removing AF points for the H?

File: src/x3f_process.c (hunk 1 above)
Comment (rolkar) [inline src/x3f_process.c]: I am not sure I understand the question. Can you give a larger code snippet and point out where you want two if statements.

**Comment by rolkar** [comment]
Exactly. This only adds sd Quattro H. The new sd Quattro firmware bug is not fixed.

I think I have to think one day or two before doing it. It is not obvious what is the best choice. Some refactoring is needed. The problem is that the code assumes that the area to check for black point is defined in a rectangle in the meta data. You only send in a name to the rectangle. But, the rectangles to the left and right are defined as column ranges, as they go from top to bottom. Moreover, both ranges are stored in the same meta data. A total mismatch.

The trick is to add this without duplicating complicated code.

**Comment by mmroden** [comment]
Then I'd say this works for me, and can be merged.

File: src/x3f_process.c (hunk 1 above)
Comment (erikrk) [inline src/x3f_process.c]: I think it refers it testing with || and using the ? operator subsequently. I would rather set g to NULL first, then test for CAMERAID and colors, and finally test on p and if it's not NULL apply the correction.

Not testing the return value of x3f_get_camf_unsigned(x3f, "CAMERAID", &cameraid) is also a bug. CAMERAID can be absent.

File: src/x3f_process.c [hunk 2]
```diff
@@ -364,14 +365,19 @@ static void interpolate_bad_pixels(x3f_t *x3f, x3f_area16_t *image, int colors)
 		     bpf23[i], row,
 		     image->columns, image->rows); i++;}
 
-  /* Interpolate over autofocus pixels for sd Quattro.
+  /* Interpolate over autofocus pixels for sd Quattro an sd Quattro H.
```
Comment (erikrk) [inline src/x3f_process.c]: Typo: *and sd Quattro H.

**Comment by rolkar** [comment]
I will look over the comments and make appropriate changes.

The major thing is that it is hard coded though. Sux!

I wonder what happens if the camera saves a medium size RAW.

**Comment by erikrk** [comment]
Medium size should work fine, since only the top layer is downsampled, and
xxx_af_chroma will then be used.

Den 2 maj 2017 13:27 skrev "Roland Karlsson" <notifications@github.com>:

**Pull request merged**